fetch: add config and auth for #194

krnowak opened the pull request:
This adds JSON config in /etc/rkt/config.json and options for basic auth and oauth bearer token auth.

Credentials are sent only over HTTPS.

See #194.

kelseyhightower commented:
@krnowak This looks awesome. Can you rebase this on master so we can start the review?

kelseyhightower commented:
@krnowak I think we will need to also make some updates on the spec to either call out which methods of auth are possible and that's it's optional.

kelseyhightower commented:
I've issued a PR against the spec regarding auth on discovery endpoints: https://github.com/appc/spec/pull/115

krnowak commented:
@kelseyhightower I had an assumption that discovery does not require auth. I suppose that we can add that on top of my work. It will need some thought. A scenario for example:

In config we have:
 {
     "auth":
     {
         "coreos.com":
         {
             "type": "oauth",
             "token": "foo"
         }
     }
 }

"rkt fetch coreos.com/etcd", it tries to download image from https://coreos.com/etcd.aci, but it runs meta-discovery because there is no such file. Discovery tells us that the file resides in images.coreos.com/aci/etcd.aci. So, images.coreos.com does not fit the "coreos.com" from configuration, so no credentials are sent. We could be clever and reuse the credentials used for coreos.com/etcd. What about the wicked case when user needs to provide different credentials for coreos.com and different for images.coreos.com? Would we support such a thing?

philips commented:
I don't want to copy the way that docker does this with a single configuration file since it makes it hard to lay down new files. Can we move to a directory based configuration instead? Something like how the KeyStore works: https://github.com/coreos/rocket/blob/master/pkg/keystore/keystore.go and https://github.com/coreos/rocket/issues/215

philips commented:
```
{ 
   "rktVersion": "v1",
   "rktKind": "auth",
   "domains": [
       "coreos.com"
   ] 
   "type": "oauth",
    "token": "foo"
}
```

```
/etc/rkt/auth.d/
/usr/lib/rkt/auth.d/
```

Open questions:

- Is it desirable to have an auth type for domain wildcards?
- Do we want subdomain wildcards?

jonboulle commented:
@philips so obviously with that approach there is no user-level granularity, it's just all-or-none (e.g. `/etc/rkt` only readable by a `rocket` Unix group). Are you seeing this the same as https://github.com/coreos/rocket/issues/43 ? A non-privileged user on a given system either _can_ or _cannot_ fetch/read all images. (I would argue this is a simplification we should go with for now, but want to be explicit)

philips commented:
@jonboulle Yes, a "rkt group" belonging can fetch/read all images. They are trusted for all local operations around the store.

jonboulle commented:
OK, would like to see this ticket start moving again :-)

jonboulle commented:
@krnowak can you pick this up again please?

krnowak commented:
Sure, I'll pick it up once I'm done with cmake builder. Should be next week then.

jonboulle commented:
Great, thanks
Am 19.03.2015 09:45 schrieb "Krzesimir Nowak" <notifications@github.com>:

> Sure, I'll pick it up once I'm done with cmake builder. Should be next
> week then.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/rocket/pull/364#issuecomment-83656657>.
>

philips commented:
@krnowak Is there ongoing work for this? It is currently set for v0.5.0 which is due to release this week.

krnowak commented:
@philips: Please postpone it to v0.6.0 if that's possible. I'll start working on this either tomorrow or a day later.

krnowak commented:
This PR depends on PR #729.

krnowak commented:
BTW, if you want to test it on your own and you don't have any ACI service requiring authorization, then there is https://github.com/endocode/test-aci-auth-server/

jonboulle commented on Documentation/config-format.md:
```diff
@@ -0,0 +1,37 @@
+# Rocket Configuration File
```
Please do a global s/[Rr]ocket/rkt/

krnowak replied:
Oh, right, I forgot about it.

jonboulle commented on rkt/prepare.go:
```diff
@@ -91,9 +92,15 @@ func runPrepare(args []string) (exit int) {
 
+	config, err := config.GetConfig()
+	if err != nil {
+		stderr("cannot get configuration: %v\n", err)
```
"prepare: cannot get configuration: %v" (newline is appended by `stderr()`)

krnowak replied:
Alright.

jonboulle commented on rkt/images.go:
```diff
@@ -342,8 +345,35 @@ func (f *fetcher) downloadSignatureFile(sigurl string, out writeSyncer) error {
+	if err != nil {
+		return err
+	}
+	u, err := url.Parse(aUrl)
```
You can use req.URL rather than parsing it again, NewRequest will handle this failure for you

krnowak replied:
Oh, thanks for the hint. So I can avoid renaming the url parameter as well.

jonboulle commented on rkt/fetch_test.go:
```diff
@@ -152,19 +253,67 @@ func TestDownloading(t *testing.T) {
-		body   []byte
-		hit    bool
+		ACIURL   string
+		SigURL   string
```
SigURL and body are constant so might as well remove them from being parameterised

krnowak replied:
Yeah, I was thinking about it when I was working with this code. In the end I decided against it, because maybe somebody sometime will want to test fetching code with signatures. But that probably can be done by another test case.

jonboulle commented on rkt/config/config.go:
```diff
@@ -0,0 +1,246 @@
+// Copyright 2015 CoreOS, Inc.
```
what's your rationale for `rkt/config/` vs. just `config/` for this package?

krnowak replied:
That the config will be used only by frontend (the code inside rkt), so it does not need to be generally useful, but only useful for specific needs of a frontend.

jonboulle replied:
ok, sounds reasonable

jonboulle commented on rkt/config/config.go:
```diff
@@ -0,0 +1,246 @@
+}
+
+func GetConfig() (*Config, error) {
+	return GetConfigFrom("/usr/lib/rkt", "/etc/rkt")
```
can you const these dirs?

krnowak replied:
Sure.

jonboulle commented on rkt/config/config.go:
```diff
@@ -0,0 +1,246 @@
+	return a
+}
+
+func GetConfig() (*Config, error) {
```
Please docstring all exported functions

krnowak replied:
Will do.

jonboulle commented on rkt/config/config.go:
```diff
@@ -0,0 +1,246 @@
+		return err
+	}
+	if len(header.RktKind) == 0 {
+		return fmt.Errorf("No rktKind specified in %q", path)
```
start errors lowercase in here

jonboulle replied:
and everywhere else, actually

krnowak replied:
Ok.

jonboulle commented on rkt/config/config_test.go:
```diff
@@ -0,0 +1,204 @@
+
+const tstprefix = "config-test"
+
+func tmpConfigFile(prefix string) (*os.File, error) {
```
what's wrong with ioutil.TempFile?

krnowak replied:
Exactly, there's no way to specify suffix in ioutil.TempFile. And creating a file with TempFile and then renaming it sounded ugly to me.

I'll document it.

jonboulle commented:
This looks great! Could you please add a usage example to the documentation too?

krnowak commented:
Ok, I rebased against current master (there were some minor conflicts because of renaming `cas` to `store`), fixed the issues mentioned above. On top of that I added more checks to `addParser` and `registerSubDir` in `config.go` and unexported the `Headerer` implementations in `auth.go` - they shouldn't be public anyway.

@jonboulle wrote:
> This looks great! Could you please add a usage example to the documentation too?

Ah, I'm not sure I understand. Do you mean describing an example setup? Like contents of config files in vendor and custom dirs, how overriding vendor config with custom config works and then what might be the result of rkt invocation? Because nothing else changes in running rkt - no new parameters or commands.

krnowak commented:
Ok, I rewrote the documentation (now in Documentation/configuration.md). Hopefully this one is detailed enough and provides enough examples to be clear how to actually use it.

jonboulle commented on rkt/config/config.go:
```diff
@@ -0,0 +1,276 @@
+		panic("kinds array cannot be empty when registering config subdir")
+	}
+	allKinds := toArray(toSet(append(configSubDirs[dir], kinds...)))
+	sort.Strings(allKinds)
```
redundant with 101 fwiw, could just define that func as returning sorted results (or really inline all that here since it isn't used elsewhere yet)

krnowak replied:
I simply forgot to remove line 101. Will do it in a moment. But I'd rather avoid inlining `toSet` and `toArray` into `registerSubDir` as they are lower-level stuff than `registerSubDir` and could litter that function with unnecessary details.

jonboulle commented:
This looks great to me. I think the one caveat is that we haven't really answered either of the open questions in https://github.com/coreos/rkt/pull/364#issuecomment-69989710 , but I think we can just do that in follow up.

jonboulle commented:
LGTM

jonboulle commented:
@philips landing, please circle back when you get a chance and we can roll forward
